### PR TITLE
Create side comments cache table

### PR DESCRIPTION
**Pull request by oetherington**
NB: If you want to run this migration on a shared dev database you should comment out the `dropField` line first to avoid breaking the DB for anybody else using it. The code will still work fine - the "removed" field will be shadowed by a SQL resolver field.

The posts table currently contains a field called `sideCommentsCache` which caches the data needed to display side comments. This includes a copy of the entire post body as HTML, which means it roughly doubles the byte footprint of each post row (this approximation is more true the longer the post). This PR removes this column and adds a separate table to store this data. There are several semi-large architectural changes needed here which may or may not stand the test of time.

Currently it works like this:
 1. Client visits a posts page which does `SELECT * FROM "Posts" WHERE ...` including the side comments cache, but the posts page fragment doesn't actually include the side comments so they're discarded before being sent to the client (but we still had to download the data from Postgres to the server).
 2. The `PostBody` component runs a `useSingle` hook requesting the `PostsSideComments` fragment which just contains the `sideComments` field. This is a resolver only field which fetches the post to the server with another `SELECT * FROM "Posts" WHERE ...`.
 3. Inside the resolver we fetch _all_ the comments for the post with `SELECT * FROM "Comments" WHERE ...`. This is because we need to sort the comments into different groups based on their author and karma.
 4. We send the required data to the client.

With this PR the process becomes:
 1. Client visits the posts page which still runs `SELECT * FROM "Posts" WHERE ...`, but now the object is about 50% smaller.
 2. `PostBody` requests the `PostsSideComments` fragment which uses a SQL resolver to fetch the post along with the cache data via a join.
 3. We still have to fetch comments in the `sideComments` resolver, but the logic has been improved:
   a. If the cache is valid then we only need to fetch the comments directly referenced by the cache, and we can add a projection to only fetch the `_id`, `userId` and `baseScore`.
   b. If the cache is invalid then we need to fetch all the comments, but only the fields above plus the `contents`.
 4. We send the required data to the client.

Aside from this, there are two places which should also see good performance boosts:
 - Any pages where we display posts lists (such as the frontpage) where we need to fetch a lot of posts but we never need to display side comments.
 - Forums that don't enable side comments (like the EA forum) now completely disable the resolver so the side comments cache should never be used.

To test the side comments functionality which is disabled for the EA Forum, CEA devs will need to set:
```
    "comments": {
         "sideCommentsEnabled": true
    },
```
in the dev public instance settings.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206595651690014) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/sideCommentCaches/collection.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import { getDefaultMutations } from "../../vulcan-core/default_mutations";
+import { ensureIndex } from "../../collectionIndexUtils";
+import schema from "./schema";
+
+export const SideCommentCaches: SideCommentCachesCollection = createCollection({
+  collectionName: "SideCommentCaches",
+  typeName: "SideCommentCache",
+  schema,
+  resolvers: getDefaultResolvers("SideCommentCaches"),
+  mutations: getDefaultMutations("SideCommentCaches"),
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/sideCommentCaches/collection.ts:12]: Possibly we should add collection-level admin-gating, in addition to denying permissions to all individual fields?

File: packages/lesswrong/lib/collections/posts/fragments.ts [hunk 2]
```diff
@@ -496,6 +496,7 @@ registerFragment(`
 registerFragment(`
   fragment PostsEdit on Post {
     ...PostsDetails
+    ...PostSideComments
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/fragments.ts:499]: Seems fine, but I admit to being confused

File: packages/lesswrong/lib/utils/schemaUtils.ts [hunk 3]
```diff
@@ -212,6 +212,7 @@ export const simplSchemaToGraphQLtype = (type: any): string|null => {
 interface ResolverOnlyFieldArgs<N extends CollectionNameString> extends CollectionFieldSpecification<N> {
   resolver: (doc: ObjectsByCollectionName[N], args: any, context: ResolverContext) => any,
   sqlResolver?: SqlResolver<N>,
+  sqlPostProcess?: SqlPostProcess<N>,
```
Comment (jpaddison3) [inline packages/lesswrong/lib/utils/schemaUtils.ts:215]: And/or docstrings here

File: packages/lesswrong/lib/vulcan-core/default_resolvers.ts [hunk 4]
```diff
@@ -208,16 +208,18 @@ export function getDefaultResolvers<N extends CollectionNameString>(
...
+          delete selector.documentId;
+        }
+        const documentId = selector._id;
 
         // get fragment from GraphQL AST
-        const fragmentName = getFragmentNameFromInfo(info, "results");
+        const fragmentName = getFragmentNameFromInfo(info, "result");
 
         let doc: ObjectsByCollectionName[N] | null;
-        if (documentId) {
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/lib/vulcan-core/default_resolvers.ts:218]: I'm a bit suss currently that this is a net positive change.

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

**Comment by jpaddison3** [comment]
Another request: I found myself confused as to the purpose of the `resolver` field within `SqlResolverJoin`? Could I request a docstring in `schemaTypes`?

File: packages/lesswrong/server/resolvers/postResolvers.ts [hunk 5]
```diff
@@ -134,42 +136,78 @@ augmentFieldsDict(Posts, {
...
+        // If the post was fetched with a SQL resolver then we will already
+        // have the side comments cache available (even though the type system
+        // doesn't know about it), otherwise we have to fetch it from the DB.
+        const sqlFetchedPost = post as unknown as PostSideComments;
+        const cache = sqlFetchedPost.sideCommentsCache === undefined
+            ? await SideCommentCaches.findOne({
+              postId: post._id,
+              version: sideCommentCacheVersion,
+            })
+            : sqlFetchedPost.sideCommentsCache;
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/server/resolvers/postResolvers.ts:155]: This is incorrect. I added a doc comment to explain why.

File: packages/lesswrong/lib/collections/posts/fragments.ts [hunk 6]
```diff
@@ -663,10 +663,27 @@ registerFragment(`
...
+ * the code resolver for `sideComments`.
+ *
+ * The order of the fields is very important. The cache is permission gated to
+ * prevent it from being sent to the client, but it needs to be accessible to
+ * the code resolver for `sideComments`. GraphQL resolves the fields _in the order_
+ * that they are defined in the fragment. The cache must be specified after the
+ * main field otherwise it will be removed by its permission gate. (There's no
+ * sensitive data in the cache so technically this isn't the end of the word, but
+ * it is a _big_ field that we don't want to waste bandwidth on).
+ */
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/lib/collections/posts/fragments.ts:680]: I tried adding a `dependsOn`, but it's more complicated than expected. In particular, it requires making some pretty intrusive changes to the way that permissions work which I think aren't worth doing unless we can find another field that needs to make use of it. My vote is that we leave it as is, but then consider it again if we find another such field.

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 7]
```diff
@@ -2469,17 +2471,42 @@ const schema: SchemaType<"Posts"> = {
...
+   * permissions here allow anybody to read this field (which is needed to
+   * make this data accessible in the resolver) but the sqlPostProcess function
+   * always sets the result to null to avoid sending large amounts of duplicated
+   * data to the client (the data isn't sensitive though - just large).
+   */
+  sideCommentsCache: resolverOnlyField({
+    type: "SideCommentCache",
+    graphQLtype: "SideCommentCache",
+    canRead: ["guests"],
+    resolver: ({_id}: DbPost) => {
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/lib/collections/posts/schema.tsx:2488]: Yes, for resolvers that aren't `useSingle` or `useMulti`, but hopefully it shouldn't be needed much longer.

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 8]
```diff
@@ -2469,17 +2471,42 @@ const schema: SchemaType<"Posts"> = {
...
+      sqlResolver: ({field, join}) => join({
+        table: "SideCommentCaches",
+        type: "left",
+        on: {
+          postId: field("_id"),
+          version: `${sideCommentCacheVersion}`,
+        },
+        resolver: (sideCommentsField) => sideCommentsField("*"),
+      }),
+      sqlPostProcess: () => null,
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/lib/collections/posts/schema.tsx:2507]: There's a relevant doc comment above on the field

File: packages/lesswrong/lib/types/schemaTypes.ts [hunk 9]
```diff
@@ -51,6 +51,12 @@ type SqlResolverArgs<N extends CollectionNameString> = {
 
 type SqlResolver<N extends CollectionNameString> = (args: SqlResolverArgs<N>) => string;
 
+type SqlPostProcess<N extends CollectionNameString> = (
+  value: AnyBecauseHard,
+  root: ObjectsByCollectionName[N],
+  context: ResolverContext,
+) => AnyBecauseHard;
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/lib/types/schemaTypes.ts:58]: I'm not sure what you mean here?

**Pull request merged**